tools/swift-stdlib-tool/swift-stdlib-tool.cpp

```diff
@@ -670,583 +643,576 @@ -(NSString *)sst_command {
 
 
 // This executable's own path.
-NSString *self_executable = []() -> NSString * {
+std::filesystem::path self_executable = []() -> std::filesystem::path {
     uint32_t len = 0;
-    _NSGetExecutablePath(nil, &len);
+    _NSGetExecutablePath(NULL, &len);
     std::vector<char> buffer;
     buffer.reserve(len);
     _NSGetExecutablePath(buffer.data(), &len);
-    return [[NSString alloc] initWithUTF8String:buffer.data()];
+    return buffer.data();
 }();
 
 
 // This executable's own xctoolchain path.
-NSString *self_toolchain = []() -> NSString * {
-    @autoreleasepool {
-        NSString *result = self_executable;
-        
-        // Remove the executable name.
-        result = result.stringByDeletingLastPathComponent;
-        
-        // Remove trailing /usr/bin, if any
-        if ([result.lastPathComponent isEqualToString:@"bin"]) {
-            result = result.stringByDeletingLastPathComponent;
-        }
-        if ([result.lastPathComponent isEqualToString:@"usr"]) {
-            result = result.stringByDeletingLastPathComponent;
-        }
-        return [result retain];
+std::filesystem::path self_toolchain = []() -> std::filesystem::path  {
+    auto result = self_executable;
+    
+    // Remove the executable name.
+    result = result.parent_path();
+    
+    // Remove trailing /usr/bin, if any
+    if (result.filename() == std::filesystem::path("bin")) {
+        result = result.parent_path();
     }
+    if (result.filename() == std::filesystem::path("usr")) {
+        result = result.parent_path();
+    }
+    return result;
 }();
 
+std::vector<uint8_t> readToEOF(int fd) {
```

Comment:
I think we can replace this function with `dispatch_read` in a loop.

Reply:
I'm also using this function to read synchronously on line 741 so I think this is probably still needed.